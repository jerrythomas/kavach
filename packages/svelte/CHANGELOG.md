# @kavach/svelte

## 1.0.0-next.6

### Patch Changes

- Added store for auth status and used alerts to display status on demo page.
- Updated dependencies
  - @kavach/cookie@1.0.0-next.6
  - @kavach/core@1.0.0-next.6
  - @kavach/deflector@1.0.0-next.6
  - @kavach/logger@1.0.0-next.6

## 1.0.0-next.5

### Patch Changes

- chore: release 1.0.0-next.4
- Split core package into logger and deflector
- 1a35ad0: Resolved linter errors.
- Releasing the next version.
- Updated dependencies
- Updated dependencies
- Updated dependencies [1a35ad0]
- Updated dependencies
  - @kavach/cookie@1.0.0-next.5
  - @kavach/core@1.0.0-next.5
  - @kavach/deflector@1.0.0-next.5
  - @kavach/logger@1.0.0-next.5
