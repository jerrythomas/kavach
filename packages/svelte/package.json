{
	"name": "@kavach/svelte",
	"version": "1.0.0-next.2",
	"description": "Drop in security for web apps.",
	"author": "Jerry Thomas <me@jerrythomas.name>",
	"license": "MIT",
	"main": "index.js",
	"svelte": "src/index.js",
	"module": "src/index.js",
	"types": "dist/index.d.ts",
	"type": "module",
	"repository": {
		"type": "git",
		"url": "https://github.com/jerrythomas/kavach",
		"directory": "packages/kavach"
	},
	"prettier": "@jerrythomas/prettier-config",
	"publishConfig": {
		"access": "public"
	},
	"scripts": {
		"prepublishOnly": "tsc",
		"lint": "prettier --check --plugin-search-dir=. . && eslint .",
		"format": "prettier --write --plugin-search-dir=. .",
		"test:ci": "vitest run",
		"test:ui": "vitest --ui",
		"test": "vitest",
		"coverage": "vitest run --coverage",
		"upgrade": "pnpm upgrade"
	},
	"devDependencies": {
		"@jerrythomas/prettier-config": "^1.0.0",
		"@sveltejs/vite-plugin-svelte": "^2.0.2",
		"@testing-library/svelte": "^3.2.2",
		"@vitest/ui": "~0.12.10",
		"c8": "^7.12.0",
		"eslint-config-svelte": "workspace:*",
		"js-yaml": "^4.1.0",
		"jsdom": "^19.0.0",
		"shared-config": "workspace:^1.0.0",
		"svelte": "^3.55.0",
		"typescript": "^4.9.4",
		"vite": "^4.0.3",
		"vitest": "~0.19.1"
	},
	"exports": {
		"./package.json": "./package.json",
		".": {
			"types": "./types/index.d.ts",
			"import": "./src/index.js"
		}
	},
	"dependencies": {
		"@kavach/cookie": "workspace:1.0.0-next.2",
		"@kavach/core": "workspace:*",
		"@svelte-spice/form": "1.0.0-next.3",
		"@svelte-spice/input": "1.0.0-next.3",
		"ramda": "^0.28.0"
	}
}
