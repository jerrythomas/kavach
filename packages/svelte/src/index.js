export * from './kavach'
export * from './request'
export * from './endpoints'

export { default as AuthButton } from './AuthButton.svelte'
export { default as AuthProvider } from './AuthProvider.svelte'
export { default as AuthGroup } from './AuthGroup.svelte'
export { default as AuthError } from './AuthError.svelte'
export { default as AuthResponse } from './AuthResponse.svelte'
export { default as AuthHandler } from './AuthHandler.svelte'
export { default as Logout } from './Logout.svelte'
