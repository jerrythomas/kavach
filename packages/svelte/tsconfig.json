{
  "extends": "shared-config/jsconfig.json",
  "include": ["src/**/*.js", "src/**/*.svelte"],
  "compilerOptions": {
    "outDir": "dist"
  }
}
