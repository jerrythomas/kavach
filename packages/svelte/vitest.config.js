import { defineConfig } from 'vite'
import { svelte } from '@sveltejs/vite-plugin-svelte'
import { testConfig } from 'shared-config'

testConfig.coverage.provider = 'v8'
export default defineConfig({
	plugins: [svelte({ hot: !process.env.VITEST })],
	test: testConfig
})
