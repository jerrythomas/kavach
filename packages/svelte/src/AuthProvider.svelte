<script>
	import { IconButton, InputField } from '@svelte-spice/form'
	import { getContext } from 'svelte'
	import { createEventDispatcher } from 'svelte'

	const dispatch = createEventDispatcher()
	const kavach = getContext('kavach')

	/** @type {'otp'|'oauth'|'password'} */
	export let mode = 'oauth'
	/** @type {string} */
	export let name
	/** @type {string} */
	export let label
	/** @type {Array<string>} */
	export let scopes = []

	let result
	let value
	let password
	async function signIn() {
		if (mode === 'password') {
			result = await kavach.signIn({ [name]: value, password })
		} else if (mode === 'otp') {
			result = await kavach.signIn({ provider: name, email: value })
		} else {
			await kavach.signIn({ provider: name, scopes })
		}
		if (result) {
			if (result.error) {
				dispatch('error', result.error)
			} else {
				dispatch('success', result.data)
			}
		}
	}
</script>

{#if mode === 'oauth'}
	<IconButton on:click={signIn} {label} leftIcon="logo-{name}" />
	<!-- <AuthButton provider={name} {label} {scopes} {params} on:click={signIn} /> -->
{:else if mode === 'password'}
	<InputField type={name} icon="logo-{name}" label={name} bind:value />
	<InputField
		type="password"
		icon="logo-password"
		label="Password"
		bind:password
	/>
	<button
		on:click={signIn}
		class="col-start-3 text-center h-10 mt-4 bg-primary text-white"
	>
		Sign in
	</button>
	<!-- <AuthPassword {label} {} on:click={signIn} /> -->
{:else}
	<!-- <pre>
		provider:{name}
		{JSON.stringify(result, null, 2)}
	</pre> -->
	<form on:submit={signIn}>
		<InputField
			type="email"
			name="email"
			icon="logo-{name}"
			placeholder={label}
			bind:value
		/>
	</form>
{/if}
