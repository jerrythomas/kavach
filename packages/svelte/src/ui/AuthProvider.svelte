<script>
	import { Button } from '@rokkit/molecules'
	import { InputField } from '@rokkit/organisms'
	import { getContext } from 'svelte'
	import { createEventDispatcher } from 'svelte'
	import AuthPassword from './AuthPassword.svelte'

	const dispatch = createEventDispatcher()
	const kavach = getContext('kavach')

	let className = ''
	export { className as class }
	/** @type {'otp'|'oauth'|'password'} */
	export let mode = 'oauth'
	/** @type {string} */
	export let name
	/** @type {string} */
	export let label
	/** @type {Array<string>} */
	export let scopes = []

	let result
	let value
	let password

	async function signIn() {
		if (mode === 'password') {
			result = await kavach.signIn({ [name]: value, password })
		} else if (mode === 'otp') {
			result = await kavach.signIn({ provider: name, email: value })
		} else {
			await kavach.signIn({
				provider: name,
				scopes
			})
		}
		if (result) {
			if (result.error) {
				dispatch('error', result.error)
			} else {
				dispatch('success', result.data)
			}
		}
	}
</script>

{#if mode === 'oauth'}
	<Button
		on:click={signIn}
		{label}
		leftIcon="i-auth-{name}"
		class={className}
	/>
{:else if mode === 'password'}
	<AuthPassword bind:value bind:password on:click={signIn} />
{:else}
	<form on:submit={signIn} class="flex w-full p-0 {className}">
		<InputField
			type="email"
			name="magic"
			label="magic link"
			icon="i-auth-{name}"
			placeholder={label}
			bind:value
		/>
	</form>
{/if}
