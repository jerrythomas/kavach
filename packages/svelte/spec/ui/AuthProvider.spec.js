import { describe, expect, it, beforeEach, vi } from 'vitest'
import { cleanup, fireEvent, render } from '@testing-library/svelte'
import { toHaveBeenDispatchedWith } from './validators'
import { tick } from 'svelte'
import AuthProvider from '../../src/ui/AuthProvider.svelte'

expect.extend({ toHaveBeenDispatchedWith })
describe('AuthProvider.svelte', () => {
	const handlers = {
		error: vi.fn(),
		success: vi.fn()
	}
	const kavach = {
		signIn: vi.fn().mockImplementation((input) => {
			const result =
				input.email === 'john.doe@example.com'
					? { data: { token: 'token' } }
					: { error: 'Invalid email' }
			return Promise.resolve(result)
		})
	}
	beforeEach(() => {
		cleanup()
		handlers.error.mockClear()
		handlers.success.mockClear()
	})

	describe('oauth', () => {
		it('should render oauth button', async () => {
			const { container, component } = render(AuthProvider, {
				context: new Map([['kavach', kavach]]),
				props: { name: 'google', label: 'Sign in with Google' }
			})
			component.$on('error', handlers.error)
			component.$on('success', handlers.success)

			const button = container.querySelector('button')
			expect(container).toMatchSnapshot()

			await fireEvent.click(button)
			expect(kavach.signIn).toHaveBeenCalledWith({
				provider: 'google',
				scopes: []
			})
			expect(handlers.success).not.toHaveBeenCalled()
			expect(handlers.error).not.toHaveBeenCalled()
		})

		it('should render oauth button with scopes', async () => {
			const { container, component } = render(AuthProvider, {
				context: new Map([['kavach', kavach]]),
				props: { name: 'azure', label: 'Sign in with Azure', scopes: ['email'] }
			})
			component.$on('error', handlers.error)
			component.$on('success', handlers.success)

			const button = container.querySelector('button')
			expect(container).toMatchSnapshot()

			await fireEvent.click(button)
			expect(kavach.signIn).toHaveBeenCalledWith({
				provider: 'azure',
				scopes: ['email']
			})
			expect(handlers.success).not.toHaveBeenCalled()
			expect(handlers.error).not.toHaveBeenCalled()
		})
	})

	describe('otp', () => {
		it('should render magic link', async () => {
			const { container, component } = render(AuthProvider, {
				context: new Map([['kavach', kavach]]),
				props: {
					mode: 'otp',
					name: 'magic',
					label: 'Sign in with Magic Link',
					class: 'custom'
				}
			})
			component.$on('error', handlers.error)
			component.$on('success', handlers.success)

			expect(container).toMatchSnapshot()
			const form = container.querySelector('form')

			await fireEvent.input(form.querySelector('input'), {
				target: { value: 'john.doe@example.com' }
			})
			await fireEvent.submit(form)
			await tick()
			expect(kavach.signIn).toHaveBeenCalledWith({
				provider: 'magic',
				email: 'john.doe@example.com'
			})
			expect(handlers.success).toHaveBeenDispatchedWith({ token: 'token' })
			expect(handlers.error).not.toHaveBeenCalled()
		})

		it('should render magic link wth custom class', async () => {
			const { container, component } = render(AuthProvider, {
				context: new Map([['kavach', kavach]]),
				props: {
					mode: 'otp',
					name: 'magic',
					label: 'Sign in with Magic Link'
				}
			})
			expect(container).toMatchSnapshot()
			component.$set({ class: 'custom', name: 'otp', label: 'Send me a link' })
			await tick()
			expect(container).toMatchSnapshot()
		})
		it('should handle errors', async () => {
			const { container, component } = render(AuthProvider, {
				context: new Map([['kavach', kavach]]),
				props: {
					mode: 'otp',
					name: 'magic',
					label: 'Sign in with Magic Link'
				}
			})
			component.$on('error', handlers.error)
			component.$on('success', handlers.success)

			expect(container).toMatchSnapshot()
			const form = container.querySelector('form')

			await fireEvent.submit(form)
			await tick()
			expect(kavach.signIn).toHaveBeenCalledWith({
				provider: 'magic',
				email: null
			})
			expect(handlers.success).not.toHaveBeenCalled()
			expect(handlers.error).toHaveBeenDispatchedWith('Invalid email')
		})
	})

	describe('password', () => {
		it('should render email/password auth', async () => {
			const { container, component } = render(AuthProvider, {
				context: new Map([['kavach', kavach]]),
				props: { mode: 'password', name: 'email', label: 'Sign in with Email' }
			})
			component.$on('error', handlers.error)
			component.$on('success', handlers.success)

			const button = container.querySelector('button')
			expect(container).toMatchSnapshot()

			await fireEvent.input(container.querySelector('input#email'), {
				target: { value: 'john.doe@example.com' }
			})
			await fireEvent.click(button)
			await tick()
			expect(kavach.signIn).toHaveBeenCalledWith({
				email: 'john.doe@example.com',
				password: ''
			})
			expect(handlers.success).toHaveBeenDispatchedWith({ token: 'token' })
			expect(handlers.error).not.toHaveBeenCalled()
		})

		it('should handle errors', async () => {
			const { container, component } = render(AuthProvider, {
				context: new Map([['kavach', kavach]]),
				props: { mode: 'password', name: 'email', label: 'Sign in with Email' }
			})
			component.$on('error', handlers.error)
			component.$on('success', handlers.success)

			const button = container.querySelector('button')
			expect(container).toMatchSnapshot()
			await fireEvent.click(button)
			await tick()
			expect(kavach.signIn).toHaveBeenCalledWith({
				email: '',
				password: ''
			})
			expect(handlers.success).not.toHaveBeenCalled()
			expect(handlers.error).toHaveBeenCalled()
			expect(handlers.error).toHaveBeenDispatchedWith('Invalid email')
		})
	})
})
