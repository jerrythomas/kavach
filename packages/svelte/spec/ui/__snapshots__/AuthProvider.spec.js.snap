// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AuthProvider.svelte > oauth > should render oauth button 1`] = `
<div>
  <button
    class="flex flex-row items-center "
    type="submit"
  >
    <icon-left
      class="flex flex-row"
    >
      <icon
        aria-label="i-auth-google"
        class="flex flex-shrink-0 items-center justify-center "
        role="img"
        tabindex="-1"
      >
        <i
          aria-hidden="true"
          class="i-auth-google"
        />
      </icon>
    </icon-left>
     
    <p>
      Sign in with Google
    </p>
    
     
  </button>
  
</div>
`;

exports[`AuthProvider.svelte > oauth > should render oauth button with scopes 1`] = `
<div>
  <button
    class="flex flex-row items-center "
    type="submit"
  >
    <icon-left
      class="flex flex-row"
    >
      <icon
        aria-label="i-auth-azure"
        class="flex flex-shrink-0 items-center justify-center "
        role="img"
        tabindex="-1"
      >
        <i
          aria-hidden="true"
          class="i-auth-azure"
        />
      </icon>
    </icon-left>
     
    <p>
      Sign in with Azure
    </p>
    
     
  </button>
  
</div>
`;

exports[`AuthProvider.svelte > otp > should handle errors 1`] = `
<div>
  <form
    class="flex w-full p-0 "
  >
    <input-field
      class="flex flex-col input-email empty"
    >
      <label
        for="magic"
      >
        magic link
      </label>
       
      <field
        aria-label="magic link"
        class="w-full flex flex-row items-center"
      >
        <icon
          aria-label="i-auth-magic"
          class="flex flex-shrink-0 items-center justify-center "
          role="img"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="i-auth-magic"
          />
        </icon>
         
         
        <input
          id="magic"
          name="magic"
          placeholder="Sign in with Magic Link"
          type="email"
        />
        
        
         
      </field>
       
    </input-field>
  </form>
  
</div>
`;

exports[`AuthProvider.svelte > otp > should render magic link 1`] = `
<div>
  <form
    class="flex w-full p-0 custom"
  >
    <input-field
      class="flex flex-col input-email empty"
    >
      <label
        for="magic"
      >
        magic link
      </label>
       
      <field
        aria-label="magic link"
        class="w-full flex flex-row items-center"
      >
        <icon
          aria-label="i-auth-magic"
          class="flex flex-shrink-0 items-center justify-center "
          role="img"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="i-auth-magic"
          />
        </icon>
         
         
        <input
          id="magic"
          name="magic"
          placeholder="Sign in with Magic Link"
          type="email"
        />
        
        
         
      </field>
       
    </input-field>
  </form>
  
</div>
`;

exports[`AuthProvider.svelte > otp > should render magic link wth custom class 1`] = `
<div>
  <form
    class="flex w-full p-0 "
  >
    <input-field
      class="flex flex-col input-email empty"
    >
      <label
        for="magic"
      >
        magic link
      </label>
       
      <field
        aria-label="magic link"
        class="w-full flex flex-row items-center"
      >
        <icon
          aria-label="i-auth-magic"
          class="flex flex-shrink-0 items-center justify-center "
          role="img"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="i-auth-magic"
          />
        </icon>
         
         
        <input
          id="magic"
          name="magic"
          placeholder="Sign in with Magic Link"
          type="email"
        />
        
        
         
      </field>
       
    </input-field>
  </form>
  
</div>
`;

exports[`AuthProvider.svelte > otp > should render magic link wth custom class 2`] = `
<div>
  <form
    class="flex w-full p-0 custom"
  >
    <input-field
      class="flex flex-col input-email empty"
    >
      <label
        for="magic"
      >
        magic link
      </label>
       
      <field
        aria-label="magic link"
        class="w-full flex flex-row items-center"
      >
        <icon
          aria-label="i-auth-magic"
          class="flex flex-shrink-0 items-center justify-center "
          role="img"
          tabindex="-1"
        >
          <i
            aria-hidden="true"
            class="i-auth-otp"
          />
        </icon>
         
         
        <input
          id="magic"
          name="magic"
          placeholder="Send me a link"
          type="email"
        />
        
        
         
      </field>
       
    </input-field>
  </form>
  
</div>
`;

exports[`AuthProvider.svelte > password > should handle errors 1`] = `
<div>
  <input-field
    class="flex flex-col input-email empty"
  >
    <label
      for="email"
    >
      email
    </label>
     
    <field
      aria-label="email"
      class="w-full flex flex-row items-center"
    >
      <icon
        aria-label="i-auth-email"
        class="flex flex-shrink-0 items-center justify-center "
        role="img"
        tabindex="-1"
      >
        <i
          aria-hidden="true"
          class="i-auth-email"
        />
      </icon>
       
       
      <input
        autocomplete="on"
        id="email"
        name="email"
        type="email"
      />
      
      
       
    </field>
     
  </input-field>
   
  <input-field
    class="flex flex-col input-password empty"
  >
    <label
      for="password"
    >
      Password
    </label>
     
    <field
      aria-label="Password"
      class="w-full flex flex-row items-center"
    >
      <icon
        aria-label="i-auth-password"
        class="flex flex-shrink-0 items-center justify-center "
        role="img"
        tabindex="-1"
      >
        <i
          aria-hidden="true"
          class="i-auth-password"
        />
      </icon>
       
       
      <input
        id="password"
        name="password"
        password=""
        type="password"
      />
      
      
       
    </field>
     
  </input-field>
   
  <button
    class="col-start-3 text-center h-10 mt-4 bg-primary text-white"
  >
    Sign in
  </button>
  
</div>
`;

exports[`AuthProvider.svelte > password > should render email/password auth 1`] = `
<div>
  <input-field
    class="flex flex-col input-email empty"
  >
    <label
      for="email"
    >
      email
    </label>
     
    <field
      aria-label="email"
      class="w-full flex flex-row items-center"
    >
      <icon
        aria-label="i-auth-email"
        class="flex flex-shrink-0 items-center justify-center "
        role="img"
        tabindex="-1"
      >
        <i
          aria-hidden="true"
          class="i-auth-email"
        />
      </icon>
       
       
      <input
        autocomplete="on"
        id="email"
        name="email"
        type="email"
      />
      
      
       
    </field>
     
  </input-field>
   
  <input-field
    class="flex flex-col input-password empty"
  >
    <label
      for="password"
    >
      Password
    </label>
     
    <field
      aria-label="Password"
      class="w-full flex flex-row items-center"
    >
      <icon
        aria-label="i-auth-password"
        class="flex flex-shrink-0 items-center justify-center "
        role="img"
        tabindex="-1"
      >
        <i
          aria-hidden="true"
          class="i-auth-password"
        />
      </icon>
       
       
      <input
        id="password"
        name="password"
        password=""
        type="password"
      />
      
      
       
    </field>
     
  </input-field>
   
  <button
    class="col-start-3 text-center h-10 mt-4 bg-primary text-white"
  >
    Sign in
  </button>
  
</div>
`;
